File: ios/IceBloxApp/Processing/LookalikeExpander.swift
import Foundation

enum LookalikeExpander {

    static func expand(
        _ text: String,
        charConfidences: [Float],
        slotCandidates: [[PlateOCR.SlotCandidate]] = [],
        maxVariants: Int = AppConfig.maxLookalikeVariants
    ) -> [(String, Int, Float)] { // swiftlint:disable:this large_tuple
        let slots = buildSlotLists(text: text, charConfidences: charConfidences, slotCandidates: slotCandidates)

        var totalCombinations: Int64 = 1
        for slot in slots {
            totalCombinations *= Int64(slot.count)
            if totalCombinations > Int64(maxVariants) { break }
        }

        if totalCombinations <= Int64(maxVariants) {
            return cartesianExpand(slots: slots, primaryText: text)
        } else {
            return priorityQueueExpand(slots: slots, primaryText: text, maxVariants: maxVariants)
        }
    }

    private static func buildSlotLists(
        text: String,
        charConfidences: [Float],
        slotCandidates: [[PlateOCR.SlotCandidate]]
    ) -> [[PlateOCR.SlotCandidate]] {
        let chars = Array(text)
        var result = [[PlateOCR.SlotCandidate]]()
        for i in chars.indices {
            if i < slotCandidates.count && slotCandidates[i].count > 1 {
                result.append(slotCandidates[i])
            } else {
                let conf: Float = i < charConfidences.count ? charConfidences[i] : 0
                result.append([PlateOCR.SlotCandidate(char: chars[i], probability: conf)])
            }
        }
        return result
    }

    private static func computeConfidence(slots: [[PlateOCR.SlotCandidate]], indices: [Int]) -> Float {
        let count = slots.count
        guard count > 0 else { return 0 }
        var logSum: Float = 0
        for i in 0..<count {
            logSum += log(max(slots[i][indices[i]].probability, 1e-6))
        }
        return exp(logSum / Float(count))
    }

    private static func cartesianExpand(
        slots: [[PlateOCR.SlotCandidate]],
        primaryText: String
    ) -> [(String, Int, Float)] {
        var results = [(String, Int, Float)]()
        let n = slots.count
        var indices = [Int](repeating: 0, count: n)

        while true {
            var chars = [Character]()
            chars.reserveCapacity(n)
            var subs = 0
            for i in 0..<n {
                chars.append(slots[i][indices[i]].char)
                if indices[i] != 0 { subs += 1 }
            }
            let conf = computeConfidence(slots: slots, indices: indices)
            results.append((String(chars), subs, conf))

            var pos = n - 1
            while pos >= 0 {
                indices[pos] += 1
                if indices[pos] < slots[pos].count { break }
                indices[pos] = 0
                pos -= 1
            }
            if pos < 0 { break }
        }

        results.sort { $0.2 > $1.2 }
        if let primaryIdx = results.firstIndex(where: { $0.0 == primaryText }), primaryIdx > 0 {
            let primary = results.remove(at: primaryIdx)
            results.insert(primary, at: 0)
        }
        return results
    }

    private static func priorityQueueExpand(
        slots: [[PlateOCR.SlotCandidate]],
        primaryText: String,
        maxVariants: Int
    ) -> [(String, Int, Float)] {
        let n = slots.count
        var results = [(String, Int, Float)]()
        var seen = Set<[Int]>()

        struct Entry: Comparable {
            let indices: [Int]
            let lastModified: Int
            let confidence: Float

            static func < (lhs: Entry, rhs: Entry) -> Bool {
                lhs.confidence > rhs.confidence
            }
        }

        var heap = [Entry]()

        func heapPush(_ entry: Entry) {
            heap.append(entry)
            var i = heap.count - 1
            while i > 0 {
                let parent = (i - 1) / 2
                if heap[i] < heap[parent] {
                    heap.swapAt(i, parent)
                    i = parent
                } else { break }
            }
        }

        func heapPop() -> Entry {
            let top = heap[0]
            let last = heap.removeLast()
            if !heap.isEmpty {
                heap[0] = last
                var i = 0
                while true {
                    let left = 2 * i + 1
                    let right = 2 * i + 2
                    var smallest = i
                    if left < heap.count && heap[left] < heap[smallest] { smallest = left }
                    if right < heap.count && heap[right] < heap[smallest] { smallest = right }
                    if smallest == i { break }
                    heap.swapAt(i, smallest)
                    i = smallest
                }
            }
            return top
        }

        let seedIndices = [Int](repeating: 0, count: n)
        let seedConf = computeConfidence(slots: slots, indices: seedIndices)
        heapPush(Entry(indices: seedIndices, lastModified: 0, confidence: seedConf))
        seen.insert(seedIndices)

        while !heap.isEmpty && results.count < maxVariants {
            let entry = heapPop()
            var chars = [Character]()
            chars.reserveCapacity(n)
            var subs = 0
            for i in 0..<n {
                chars.append(slots[i][entry.indices[i]].char)
                if entry.indices[i] != 0 { subs += 1 }
            }
            results.append((String(chars), subs, entry.confidence))

            for pos in entry.lastModified..<n {
                let nextIdx = entry.indices[pos] + 1
                if nextIdx < slots[pos].count {
                    var childIndices = entry.indices
                    childIndices[pos] = nextIdx
                    if seen.insert(childIndices).inserted {
                        let conf = computeConfidence(slots: slots, indices: childIndices)
                        heapPush(Entry(indices: childIndices, lastModified: pos, confidence: conf))
                    }
                }
            }
        }

        return results
    }
}

File: android/app/src/main/java/com/iceblox/app/processing/LookalikeExpander.kt
package com.iceblox.app.processing

import com.iceblox.app.config.AppConfig
import com.iceblox.app.detection.SlotCandidate
import java.util.PriorityQueue
import kotlin.math.exp
import kotlin.math.ln
import kotlin.math.max

object LookalikeExpander {

    fun expand(
        text: String,
        charConfidences: FloatArray,
        slotCandidates: List<List<SlotCandidate>> = emptyList(),
        maxVariants: Int = AppConfig.MAX_LOOKALIKE_VARIANTS
    ): List<Triple<String, Int, Float>> {
        val slots = buildSlotLists(text, charConfidences, slotCandidates)

        var totalCombinations = 1L
        for (slot in slots) {
            totalCombinations *= slot.size
            if (totalCombinations > maxVariants) break
        }

        return if (totalCombinations <= maxVariants) {
            cartesianExpand(slots, text)
        } else {
            priorityQueueExpand(slots, text, maxVariants)
        }
    }

    private fun buildSlotLists(
        text: String,
        charConfidences: FloatArray,
        slotCandidates: List<List<SlotCandidate>>
    ): List<List<SlotCandidate>> {
        val result = mutableListOf<List<SlotCandidate>>()
        for (i in text.indices) {
            if (i < slotCandidates.size && slotCandidates[i].size > 1) {
                result.add(slotCandidates[i])
            } else {
                val conf = if (i < charConfidences.size) charConfidences[i] else 0f
                result.add(listOf(SlotCandidate(text[i], conf)))
            }
        }
        return result
    }

    private fun computeConfidence(slots: List<List<SlotCandidate>>, indices: IntArray): Float {
        val count = slots.size
        if (count == 0) return 0f
        var logSum = 0f
        for (i in 0 until count) {
            logSum += ln(max(slots[i][indices[i]].probability, 1e-6f))
        }
        return exp(logSum / count)
    }

    private fun cartesianExpand(
        slots: List<List<SlotCandidate>>,
        primaryText: String
    ): List<Triple<String, Int, Float>> {
        val results = mutableListOf<Triple<String, Int, Float>>()
        val n = slots.size
        val indices = IntArray(n)

        while (true) {
            val sb = StringBuilder(n)
            var subs = 0
            for (i in 0 until n) {
                sb.append(slots[i][indices[i]].char)
                if (indices[i] != 0) subs++
            }
            val conf = computeConfidence(slots, indices)
            results.add(Triple(sb.toString(), subs, conf))

            var pos = n - 1
            while (pos >= 0) {
                indices[pos]++
                if (indices[pos] < slots[pos].size) break
                indices[pos] = 0
                pos--
            }
            if (pos < 0) break
        }

        results.sortByDescending { it.third }
        val primaryIdx = results.indexOfFirst { it.first == primaryText }
        if (primaryIdx > 0) {
            val primary = results.removeAt(primaryIdx)
            results.add(0, primary)
        }
        return results
    }

    private fun priorityQueueExpand(
        slots: List<List<SlotCandidate>>,
        primaryText: String,
        maxVariants: Int
    ): List<Triple<String, Int, Float>> {
        val n = slots.size
        val results = mutableListOf<Triple<String, Int, Float>>()
        val seen = mutableSetOf<List<Int>>()

        data class Entry(
            val indices: IntArray,
            val lastModified: Int,
            val confidence: Float
        )

        val queue = PriorityQueue<Entry>(compareByDescending { it.confidence })

        val seedIndices = IntArray(n)
        val seedConf = computeConfidence(slots, seedIndices)
        queue.add(Entry(seedIndices, 0, seedConf))
        seen.add(seedIndices.toList())

        while (queue.isNotEmpty() && results.size < maxVariants) {
            val entry = queue.poll()!!
            val sb = StringBuilder(n)
            var subs = 0
            for (i in 0 until n) {
                sb.append(slots[i][entry.indices[i]].char)
                if (entry.indices[i] != 0) subs++
            }
            results.add(Triple(sb.toString(), subs, entry.confidence))

            for (pos in entry.lastModified until n) {
                val nextIdx = entry.indices[pos] + 1
                if (nextIdx < slots[pos].size) {
                    val childIndices = entry.indices.copyOf()
                    childIndices[pos] = nextIdx
                    val key = childIndices.toList()
                    if (seen.add(key)) {
                        val conf = computeConfidence(slots, childIndices)
                        queue.add(Entry(childIndices, pos, conf))
                    }
                }
            }
        }

        return results
    }
}
